src/source/src/parser/mod.rs

```diff
@@ -88,51 +88,144 @@ pub struct SourceStreamChunkRowWriter<'a> {
 /// the [`SourceParser::parse`] implementation.
 pub struct WriteGuard(());
 
+trait OpAction {
+    type Output;
+
+    const DEFAULT_OUTPUT: Self::Output;
+
+    fn apply(builder: &mut ArrayBuilderImpl, output: Self::Output) -> Result<()>;
+
+    fn rollback(builder: &mut ArrayBuilderImpl);
+
+    fn finish(writer: SourceStreamChunkRowWriter<'_>);
+}
+
+struct OpActionInsert;
+
+impl OpAction for OpActionInsert {
+    type Output = Datum;
+
+    const DEFAULT_OUTPUT: Self::Output = None;
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: Datum) -> Result<()> {
+        builder.append_datum(&output).map_err(Into::into)
+    }
+
+    #[inline(always)]
+    fn rollback(builder: &mut ArrayBuilderImpl) {
+        builder.pop().unwrap()
+    }
+
+    #[inline(always)]
+    fn finish(writer: SourceStreamChunkRowWriter<'_>) {
+        writer.op_builder.push(Op::Insert)
+    }
+}
+
+struct OpActionDelete;
+
+impl OpAction for OpActionDelete {
+    type Output = Datum;
+
+    const DEFAULT_OUTPUT: Self::Output = None;
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: Datum) -> Result<()> {
+        builder.append_datum(&output).map_err(Into::into)
+    }
+
+    #[inline(always)]
+    fn rollback(builder: &mut ArrayBuilderImpl) {
+        builder.pop().unwrap()
+    }
+
+    #[inline(always)]
+    fn finish(writer: SourceStreamChunkRowWriter<'_>) {
+        writer.op_builder.push(Op::Delete)
+    }
+}
+
+struct OpActionUpdate;
+
+impl OpAction for OpActionUpdate {
+    type Output = (Datum, Datum);
+
+    const DEFAULT_OUTPUT: Self::Output = (None, None);
+
+    #[inline(always)]
+    fn apply(builder: &mut ArrayBuilderImpl, output: (Datum, Datum)) -> Result<()> {
+        builder.append_datum(&output.0)?;
```

Comment:
There is a bug, the `builder` may failed on the intermediate state...

I will use unwrap here